### PR TITLE
feat: add session start limit data

### LINKED ISSUE #390: Don't throw away Session Start Limit
### Summary

https://discord.com/developers/docs/topics/gateway#session-start-limit-object

### What is the feature request for?

The core library

### The Problem

unconsumed data

### The Ideal Solution

consume data, perhaps add it as a client attribute

### The Current Solution

manually fetch the endpoint with auth? No idea.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Pretty self-explanatory; adds session start limit data to the client object, which was previously just dropped.
The contained `max_concurrency` value isn't being used, but given how the shard launching process currently works it's not really necessary.

resolves #390

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
    - `HTTPClient.get_bot_gateway` return type changed
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by onerandomusername** [comment]
I think session start limit could be a named tuple

**Comment by onerandomusername** [comment]
Named tuple doesn't make much sense, after thinking about it. Good choices

File: disnake/client.py [hunk 1]
```diff
@@ -265,6 +266,11 @@ class Client:
     asyncio_debug: :class:`bool`
         Whether to enable asyncio debugging when the client starts.
         Defaults to False.
+    session_start_limit: Optional[:class:`SessionStartLimit`]
+        Information about the current session start limit,
+        only set in sharded clients/bots.
```
Comment (onerandomusername) [inline disnake/client.py:271]: This is a bit of a problem. Session start limit data is also provided when not sharded, and is still important to have. To change, we would need to use the get bot gateway endpoint for non-sharded clients as well.

**Review by onerandomusername — CHANGES_REQUESTED** [review]
A small problem with the implemenatation itself. Please ignore part of my previous comment.

File: disnake/client.py (hunk 1 above)
Comment (shiftinv) [reply disnake/client.py]: https://github.com/DisnakeDev/disnake/pull/402/commits/200d0ebe3e6abd2cf1fcddea4597287cd072e342

File: disnake/client.py [hunk 2]
```diff
@@ -140,6 +142,51 @@ def _cleanup_loop(loop: asyncio.AbstractEventLoop) -> None:
...
+        self.remaining: int = data["remaining"]
+        self.reset_after: int = data["reset_after"]
+        self.max_concurrency: int = data["max_concurrency"]
+
+        self.reset_time: datetime = utils.utcnow() + timedelta(milliseconds=self.reset_after)
+
+    def __repr__(self):
+        return (
+            f"<SessionStartLimit total={self.total!r} remaining={self.remaining!r} "
+            f"reset_after={self.reset_after!r} max_concurrency={self.max_concurrency!r} reset_time={self.reset_time!s}"
```
Comment (onerandomusername) [inline disnake/client.py:186]:
```suggestion
            f"reset_after={self.reset_after!r} max_concurrency={self.max_concurrency!r} reset_time={self.reset_time!s}">
```

**Review by onerandomusername — CHANGES_REQUESTED** [review]

**Review by onerandomusername — APPROVED** [review]
works

**Pull request merged**